rfcs/0001-syntax-tree-patterns.md

```diff
@@ -0,0 +1,797 @@
+
+## Implementing clippy lints using patterns
+
+As a "real-world" example, I re-implemented the `collapsible_if` lint using patterns. The code can be found [here](https://github.com/fkohlgrueber/rust-clippy-pattern/blob/039b07ecccaf96d6aa7504f5126720d2c9cceddd/clippy_lints/src/collapsible_if.rs#L88-L163). The pattern-based version passes all test cases that were written for `collapsible_if`.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Overview
+
+The following diagram shows the dependencies between the main parts of the proposed solution:
+
+```
+                          Pattern syntax                                     
+                                |                                           
+                                |  parsing / lowering                       
+                                v                                           
+                           PatternTree                                      
+                                ^                                             
+                                |                                             
+                                |                                             
+                          IsMatch trait                                       
+                                |                                             
+                                |                                             
+             +---------------+-----------+---------+                      
+             |               |           |         |                      
+             v               v           v         v                      
+        syntax::ast     rustc::hir      syn       ...               
+```
+
+The pattern syntax described in the previous section is parsed / lowered into the so-called *PatternTree* data structure that represents a valid syntax tree pattern. Matching a *PatternTree* against an actual syntax tree (e.g. rust ast / hir or the syn ast, ...) is done using the *IsMatch* trait.
+
+The *PatternTree* and the *IsMatch* trait are introduced in more detail in the following sections.
+
+## PatternTree
+
+The core data structure of this RFC is the **PatternTree**. 
+
+It's a data structure similar to rust's AST / HIR, but with the following differences:
+
+- The PatternTree doesn't contain parsing information like `Span`s
+- The PatternTree can represent alternatives, sequences and optionals
+
+The code below shows a simplified version of the current PatternTree:
+
+> Note: The current implementation can be found [here](https://github.com/fkohlgrueber/pattern-matching/blob/dfb3bc9fbab69cec7c91e72564a63ebaa2ede638/pattern-match/src/pattern_tree.rs#L50-L96).
+
+
+```
+pub enum Expr {
+    Lit(Alt<Lit>),
+    Array(Seq<Expr>),
+    Block_(Alt<BlockType>),
+    If(Alt<Expr>, Alt<BlockType>, Opt<Expr>),
```

Comment:
For consistency, perhaps users of opt can be made to use named parameters? I.e. 

`If(_, _)`, vs `If(_, _, then=_)` vs `If(_, _, then=_?)`. It might be cleaner that way. I'm not sure.